src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs

```diff
@@ -149,11 +154,51 @@ void Dispose(bool disposing)
             }
         }
 
-        void CheckDisposed()
+        internal void SetReadOnly()
+        {
+            this.readOnly = true;
+        }
+
+        void CheckDisposedOrReadOnly()
         {
             if (this.disposed)
             {
-                throw RelayEventSource.Log.ThrowingException(new ObjectDisposedException(this.GetType().FullName), this);
+                string message = this.Context.TrackingContext.EnsureTrackableMessage(this.ToString());
+                throw RelayEventSource.Log.ThrowingException(new ObjectDisposedException(message), this);
+            }
+
+            if (this.readOnly)
+            {
+                string message = this.Context.TrackingContext.EnsureTrackableMessage(SR.ResponseBodyStarted);
+                throw RelayEventSource.Log.ThrowingException(new InvalidOperationException(message), this);
+            }
+        }
+
+        class ResponseWebHeaderCollection : WebHeaderCollection
+        {
+            readonly RelayedHttpListenerResponse response;
+
+            public ResponseWebHeaderCollection(RelayedHttpListenerResponse response)
+            {
+                this.response = response;
+            }
+
+            public override void Add(string name, string value)
+            {
+                this.response.CheckDisposedOrReadOnly();
+                base.Add(name, value);
+            }
+
+            public override void Remove(string name)
```

Comment:
>Remove(string name) [](start = 33, length = 19)

Just curious, with override like this what happens to the other overloads in the base class? Are those inaccessible now?

Reply:
We can only intercept methods which are virtual. I've gone through the list and don't see any others which are obvious.

If the base class is well-designed then all other overloads should eventually result in one of these being invoked.

---
In reply to: [276870012](https://github.com/Azure/azure-relay-dotnet/pull/105#discussion_r276870012) [](ancestors = 276870012)

Reply:
Another example:
The base class' `public string this[string name]`
is implemented by calling
`public virtual void Set(string name, string value)`

---
In reply to: [276871458](https://github.com/Azure/azure-relay-dotnet/pull/105#discussion_r276871458) [](ancestors = 276871458,276870012)